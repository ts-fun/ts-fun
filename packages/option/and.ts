import { Option } from './option'
import { none } from './none'

/**
 * Return a `None` if `left` does not contain a value,
 * otherwise return `right`
 * @param left Option to match against
 * @param right Option to return when `left` contains a value
 * @returns Either `right` or a `None`
 */
export const and = <Left, Right = Left> (
  left: Option<Left>,
  right: Option<Right>
) => left.tag ? right : left

/**
 * Return `None` if `left` does not contain a value,
 * otherwise call `right` with the wrapped value and return the result
 * @param left Option to match against
 * @param right Function to call when `left` contains a value
 * @returns Either returning value of `right` or a `None`
 */
export const andThen = <Left, Right = Left> (
  left: Option<Left>,
  right: (value: Left) => Option<Right>
) => left.tag ? right(left.value) : none()
