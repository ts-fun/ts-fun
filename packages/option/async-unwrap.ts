import { AsyncOption } from './utils/types'
import { unwrap, unwrapOrElse } from './unwrap'

/**
 * Turn a `Some` into a `Promise` that resolves
 * or turn a `None` into a `Promise` that rejects
 * @param option Option to unwrap
 * @returns A `Promise` that may resolve contained value
 */
export const asyncUnwrap = async <Value = never>(option: AsyncOption<Value>) => unwrap(await option)

/**
 * Unwrap contained value of a `Some`,
 * or throws an error
 * @param option Option to unwrap
 * @param message Error to throw when `option` does not contain a value
 * @returns Contained value
 */
export const asyncUnwrapOrErr = <
  Value = never,
  Message = Error,
>(option: AsyncOption<Value>, message: Message) =>
  asyncUnwrapOrElse(option, () => {
    throw message
  })

/**
 * Return contained value of a `Some`
 * or default value
 * @param option Option to unwrap
 * @param def Value to return when `option` does not contain a value
 * @returns Promise that resolves contained value or `def`
 */
export const asyncUnwrapOr = <Value = never, Default = Value>(
  option: AsyncOption<Value>,
  def: Default,
) => asyncUnwrapOrElse(option, () => def)

/**
 * Return contained value of a `Some`
 * or call a function
 * @param option Option to unwrap
 * @param def Function to be called when `option` does not contain a value
 * @returns Promise that resolves to either contained value or returning value of `none`
 */
export const asyncUnwrapOrElse = async <Value = never, Default = Value>(
  option: AsyncOption<Value>,
  def: () => Default | Promise<Default>,
) => unwrapOrElse(await option, def)
