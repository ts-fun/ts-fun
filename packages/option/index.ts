export * from './option'
export * from './some'
export * from './none'
export * from './and'
export * from './or'
export * from './flatten'
export * from './async-flatten'
export * from './unwrap'
export * from './async-unwrap'
export * from './clone'
export * from './match'
export * from './map'
export * from './filter'
export * from './iter'
export * from './tuple'
