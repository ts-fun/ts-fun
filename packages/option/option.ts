import { Some, some } from './some'
import { None, none } from './none'

/**
 * Option type
 */
export type Option<Value> = Some<Value> | None

/**
 * Turn `null` or `undefined` into a `None`
 * and others into a `Some`
 * @param value Nullable value
 * @returns An `Option`
 */
export const option =
  <Value> (value: Value): MakeReturn<Value> =>
    (value === null || value === undefined ? none() : some(value)) as any

type MakeReturn<Value> =
  Value extends null ? None :
  Value extends undefined ? None :
  null extends Value ? Option<Value> :
  undefined extends Value ? Option<Value> :
  Some<Value>

export default option
