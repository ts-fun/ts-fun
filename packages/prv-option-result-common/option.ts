import { Some, None } from './types'
import { box } from './box'

/**
 * Create a `Some`
 * @param value Value to be carried
 * @returns `Some` of `value`
 */
export const some: <Value> (value: Value) => Some<Value> = box

/**
 * Create a `None`
 * @returns A `None`
 */
export const none = (): None => ({ tag: false })
