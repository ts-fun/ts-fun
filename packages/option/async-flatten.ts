import { AsyncDeepOption } from './utils/types'
import { Option } from './option'
import { some } from './some'

/**
 * Flatten `deepOptionPromise`
 * @param deepOptionPromise Option to be flatten
 * @returns Promise that resolves flattened option
 */
export async function asyncFlatten <Value = never> (
  deepOptionPromise: AsyncDeepOption<Value>
): Promise<Option<Value>> {
  const deepOption = await deepOptionPromise
  if (!deepOption.tag) return deepOption

  const flatOption = await deepOption.value
  if (!flatOption.tag) return flatOption

  return some(await flatOption.value)
}

export default asyncFlatten
