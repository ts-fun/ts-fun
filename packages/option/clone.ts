import { Option } from './option'
import { some } from './some'
import { none } from './none'
import { match } from './match'

/**
 * Clone an option
 * @param option Option to clone
 * @returns A new copy of `option`
 */
export const clone = <Value = never>(option: Option<Value>) => match(option, { some, none })

export default clone
