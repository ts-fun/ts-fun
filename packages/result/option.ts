import { Option } from '@tsfun/option'
import { Result } from './result'

/**
 * Convert `ok(x)` to an `some(x)`
 * or `err(_)` to `none()`
 * @param result Result to convert
 * @returns Corresponding `Option`
 *
 * In reality, this function merely returns the same object it takes
 * since `Result` interface is compatible with `Option`
 */
export const option = <Payload = never> (result: Result<Payload, any>): Option<Payload> => result

export default option
