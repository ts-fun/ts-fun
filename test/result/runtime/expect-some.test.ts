import {
  expectSome,
  expectSomeOr,
  expectSomeOrElse,
  ok,
  err
} from '@tsfun/result'

import { some, none } from '@tsfun/option'

describe('expectSome', () => {
  it('works with Some', () => {
    expect(expectSome(some('x'))).toEqual(ok('x'))
  })

  describe('works with None', () => {
    const result = expectSome(none())

    it('matches snapshot', () => {
      expect(result).toMatchSnapshot()
    })

    it('has expected structure', () => {
      expect(result).toMatchObject(err(expect.any(Error)))
    })
  })
})

describe('expectSomeOr', () => {
  it('works with Some', () => {
    expect(expectSomeOr(some('a'), 'b')).toEqual(ok('a'))
  })

  it('works with None', () => {
    expect(expectSomeOr(none(), 'b')).toEqual(err('b'))
  })
})

describe('expectSomeOrElse', () => {
  describe('works with Some', () => {
    const error = jest.fn(() => 'b')
    const result = expectSomeOrElse(some('a'), error)

    it('returns expected result', () => {
      expect(result).toEqual(ok('a'))
    })

    it('does not call error function', () => {
      expect(error).not.toBeCalled()
    })
  })

  describe('works with None', () => {
    const error = jest.fn(() => 'b')
    const result = expectSomeOrElse(none(), error)

    it('returns expected result', () => {
      expect(result).toEqual(err('b'))
    })

    it('calls error function exactly once', () => {
      expect(error).toBeCalledTimes(1)
    })

    it('calls error function without any argument', () => {
      expect(error).toBeCalledWith()
    })
  })
})
