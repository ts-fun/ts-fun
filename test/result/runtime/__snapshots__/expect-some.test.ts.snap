// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expectSome works with None matches snapshot 1`] = `
Object {
  "error": [Error: Expecting Some but received None],
  "tag": false,
}
`;
